pangeo_forge/storage.py

```diff
@@ -0,0 +1,65 @@
+import os
+from contextlib import contextmanager
+from dataclasses import dataclass
+
+import fsspec
+
+
+@dataclass
+class Target:
+    """Representation of a storage target for Pangeo Forge.
+    Attributes
+    ----------
+    fs : FileSystemSpec.AbtractFileSystem
+        The filesystem we are writing to. Should be instantiated outside this
+        class.
+    path : str
+        The path where the target data will be saved.
+    """
+
+    fs: fsspec.AbstractFileSystem
+    path: str
+
+    def get_mapper(self):
+        # don't want to use this because we want to use a fancier Zarr FSStore
+        return self.fs.get_mapper(self.path)
+
+
+def _hash_path(path: str) -> str:
+    return str(hash(path))
+
+
+@dataclass
+class InputCache:
+    """Representation of an intermediate storage location where remote files
+    Can be cached locally.
+
+    Attributes
+    ----------
+    fs : FileSystemSpec.AbtractFileSystem
+        The filesystem we are writing to. Should be instantiated outside this
+        class.
+    prefix : str
+        A path prepended to all paths.
+    """
+
+    fs: fsspec.AbstractFileSystem
+    prefix: str = ""
+
+    def _full_path(self, path):
+        return os.path.join(self.prefix, _hash_path(path))
```

Comment:
For what its worth, I've found that hashing paths like this can make it difficult to debug failed workflows. Maybe you can explain a bit more why we need to hash all paths like this?

Reply:
I agree. My thinking was: I want there to be unique mapping between inputs and paths in the cache. Hashing achieves this. The input paths may be urls with lots of forbidden characters in them. But I'll play with some alternatives that are more readable / debuggable.